### DISCUSSION TITLE
SVG image support?

### DISCUSSION THREAD

**Discussion by EduardoVB** [twinbasic/twinbasic#749] [Ideas]
I've been thinking about the DPI aware issue, and one of the issues is the scaling of images.
Images, specially icons, need to have different sizes for different DPI settings.

One approach is to use larger images (like 256x256) and to downscale them to the proper size according to the DPI setting. But for small icons (16x16, 24x24, 32x32) that does not produce the best results, because icons get a bit blurred.
Another approach is to have different image sizes for each DPI setting. The drawback is that you need to supply one different icon image for each DPI setting.
And another approach is to use vector based images. Fonts are vector based (FontAwesome can be used for icons), and also the SVG files (and WMF/EMF too, but those are quite old/deprecated technology). 

The idea is perhaps to support SVG files.

**Comment by WaynePhillipsEA** [comment]
Onwards of Windows 10 Creators Update, Direct2d can now allow us to render SVGs.  Without that OS support would require a library such as Cairo.

I'd rather stick with OS support for now, but I'd be interested to hear others opinions.

**Reply by WaynePhillipsEA** [reply]
One option might be to have the form designer cache a bitmap copy of the SVG, to be used as a fallback if the OS doesn't support rendering SVGs?

**Reply by EduardoVB** [reply]
> One option might be to have the form designer cache a bitmap copy of the SVG, to be used as a fallback if the OS doesn't support rendering SVGs?

Yes, may be. One key point of a vector based image is to be able to render it on any DPI (or size). I guess the fallback bitmap could be one large and then used to scale it down to the desired size.

**Reply by bclothier** [reply]
> I'd rather stick with OS support for now, but I'd be interested to hear others opinions.

Could we run into cross-plafform issues if we rely on the OS? For best cross-platform compatibility, it may be desirable to provide that internally.

**Reply by WaynePhillipsEA** [reply]
Rendering SVGs accurately is no easy task, and we really want to avoid external libraries if we can help it.

**Comment by Kr00l** [comment]
Eduardo, do you mean to support SVG images through StdPicture?
Or by a graphic helper library provided by tB? (To draw onto a HDC)

Just my very personal opinion: vbRichClient (Cairo) can be used for that. So IMO for tB a post v1 issue.

**Reply by EduardoVB** [reply]
> do you mean to support SVG images through StdPicture?

Yes (if possible), since that format is now quite popular.
And I repeat: the DPI issue is a real problem (explained in the OP)

**Reply by Kr00l** [reply]
Then we may add 32bpp icon support.
Also 32bpp pARGB bitmap support is missing for StdPicture. (pARGB being the MS standard, thus pure ARGB is neglible (not pre-multiplied) and can remain rendered w/o Alpha via IPicture:Render)

tB would need then to hack the vtable of an StdPicture instance (no global vtable subclass) to enhance the drawing.

As you see it's much to make better.

**Reply by EduardoVB** [reply]
Yes, and support for PNG and GIF is also desirable.

PD: these are ideas to consider, not a need for now.

**Reply by Kr00l** [reply]
GIF works out of box. PNG is converted to BMP by tB already?

**Reply by WaynePhillipsEA** [reply]
Yes PNG is supported in the designer (not via LoadPicture, but perhaps it should)

**Reply by WaynePhillipsEA** [reply]
I don't think we'd need to hack vtables, but instead just provide a full substitute for the StdPicture class and compatible interfaces.  Big task though.

**Reply by Kr00l** [reply]
The StdPicture object can actually hold 32bbp icons and 32bpp bitmaps. It's just the rendering which makes issues.

**Comment by EduardoVB** [comment]
Because I was thinking: tB is DPI aware, but what is it going to do with the icons? Because icons (ICO or BMP) are pixel based, they will remain in just one fixed size.

**Reply by WaynePhillipsEA** [reply]
Indeed, at the moment they will be stretched to match the DPI.

**Comment by WaynePhillipsEA** [comment]
Here's another thought.   Using Inkscape, we can convert SVGs to EMF format, and EMF format is naturally supported on Windows (and already supported in tB).  The conversion could also be done via the command line provided you've got Inkscape installed, so potentially the form designer could allow SVGs to be used, and automatically convert and persist them to EMF format.

**Reply by WaynePhillipsEA** [reply]
As of v0.14.33, provided you have Inkscape installed (https://inkscape.org/) the form designer should allow you to select SVG files, and they will be auto-converted to EMF format.  Some complex SVGs will not look perfect when converted to EMF, but from my preliminary tests, Inkscape does a very good conversion of many SVGs.

**Reply by EduardoVB** [reply]
Thank you Wayne and Krool.

I should have been more clear that it wasn't a need or a request, but an idea of how to possibly solve the icon scaling issue in the future.

I'm currently developing a new program, I started it in VB6 like two months ago, it has its own difficulties so I did in VB6 and at the time there was no forms/controls in tB.
But since I'm doing it DPI aware, the solution for icons that I use is to provide a different image for each DPI setting. In fact three icon sizes, and I manage to do all using these three sizes.
But this issue of having to make several images is something that asks for another solution, that's why I opened this thread in Discussions and as an idea, thinking on how this thing would be handled in tB (in the future).

There are many [libraries for JavaScript](https://www.webdesignerdepot.com/2018/02/8-best-free-libraries-for-svg/).
Perhaps in the future someone can study and convert the code to tB, so it can also work cross platform.

**Reply by WaynePhillipsEA** [reply]
Perhaps we could enhance the Image control to allow for an array of StdPictures, and have it choose the best match at runtime (taking into account the DPI)? Obviously, working with vectors should be preferred... but that's not always possible.

**Reply by EduardoVB** [reply]
> Perhaps we could enhance the Image control to allow for an array of StdPictures...

The image control is not very popular I think (at least I almost never use it).

The images/icons can be almost anywhere. The most common places are toolbars and command buttons.
I know it is not an easy subject, but I wanted to raise the awareness of the "problem" regarding DPI and images.

The *.ICO file format already can hold different icon sizes, but I guess once it gets into a StdPicture only one is stored.

Anyway for the developer, having to design several versions of the same icon is a pain, I do it but don't know of anyone else in the VB community that took that approach (I fact I think very few make programs DPI aware in VB6).

What you say about Image control to allow for an array of StdPictures, that's the approach I took in some of my controls, for example the [SStab replacement](https://github.com/EduardoVB/SSTabEx), I added properties Pic16, Pic20 and Pic24 that correspond to 96 DPI, 120 DPI and 144 DPI. For the other DPI settings I use one those images scaled. 
Perhaps to make it better, it would need also a Pic256 to be used for higher DPIs.

But that is a lot of images, that result in a lot of graphic work when a program has many icons in a toolbar.
Although it gives the best result, because the icons look good and crisp on any DPI setting.

The alternatives are, more or less already mentioned in the OP:
1) To use just one icon, 256x256 and scale it down to whatever size. Cons: small icons such as 16x16 and 20x20 will look blurred.
2) To use a font for icons, like FontAwesome. Cons: Icons are just monochrome, and the font need to be installed.
3) Use SVG files: Cons: not easy to implement with all features, that would include to be able to also save the stored (original) SVG to a file.
The SVG format, perhaps even being now somewhat popular, there could be still few icons in that format out there.
Also we would have to search whether there are good programs to make such icons and the difficulty (I never worked with them so far).

We can ignore the issue for now, but since tB is DPI aware by default, the problem will be more patent than with VB6.
If you scale the images automatically it is a solution, but at some point I guess the issue that "the icons look blurred" will be raised.

BTW what are your plans for DPI awareness, system aware or Per Monitor?

I know you understand it, but nevertheless I will summary how it is:
A) Unaware, like VB6 is (by default): The OS lies to the program and say it is running always on 96 DPI, regardless of the real DPI setting. Everything is virtualized and scaled by the OS (it produces windows that look a bit blurred in other DPI settings than 96 DPI).
B) System aware: The program is aware of the DPI but only of the DPI setting it starts in. If a window is moved to a different monitor or if the DPI setting is changed while the program is running, the OS (from Windows 10 and above) virtualizes the DPI setting and scales the windows automatically (the same as unaware). In my experience it works good enough.
C) Per monitor aware: The OS notifies the program of any DPI change (because of the windows was carried to another monitor or because the user changed the DPI setting from the Screen properties), the program gets the WM_DPICHANGED message (and some other messages) and the OS does not scale anything.
The program is responsible of adapting its windows to any DPI setting, needing to scale forms, controls, images and everything on the fly.

**Reply by WaynePhillipsEA** [reply]
I do agree in that I think DPI will become a common issue for tB projects (and of course migrations from VB6), so I appreciate you bringing the discussion up.

At the moment, the form engine is per-monitor DPI aware, responding to live DPI changes via WM_DPICHANGED. 
 We will be needing an option for turning this off to revert to full VB6-compatible behaviour (DPI unaware).

**Reply by Kr00l** [reply]
Isn't that option determined by the application manifest? Whether unaware, system (primary monitor) or per-monitor DPI aware?
So everyone can already influence the behavior. Or am I missing something.

Btw, I do think lots of VB6 projects target system DPI awareness.

**Reply by WaynePhillipsEA** [reply]
Yes it should be, but currently is controlled the API way.  I should probably remove that now and start doing it through the manifest.

**Comment by Kr00l** [comment]
I just want to share a helper function (which need's a IPicture as source and a hDC as target) to solve 32bpp icons and 32bpp bitmaps that could be used for the VB.Image control rendering.
It evolved with the help from [VBForums](https://www.vbforums.com/showthread.php?870723-StdPicture-GdiAlphaBlend) (especially LaVolpe/wqweto)
The wrapper can be optionally passed in ByRef with a RenderFlag which updates itself to avoid doing the same thing over and over again.

```
Public Sub RenderPicture(ByVal Picture As IPicture, ByVal hDC As Long, ByVal X As Long, ByVal Y As Long, Optional ByVal CX As Long, Optional ByVal CY As Long, Optional ByRef RenderFlag As Integer)
' RenderFlag is passed as a optional parameter ByRef.
' It is ignored for icons and metafiles.
' 0 = render method unknown, determine it and update parameter
' 1 = StdPicture.Render
' 2 = GdiAlphaBlend
If Picture Is Nothing Then Exit Sub
With Picture
If .Handle <> 0 Then
    If CX = 0 Then CX = CHimetricToPixel_X(.Width)
    If CY = 0 Then CY = CHimetricToPixel_Y(.Height)
    If .Type = vbPicTypeIcon Then
        Const DI_NORMAL As Long = &H3
        DrawIconEx hDC, X, Y, .Handle, CX, CY, 0, 0, DI_NORMAL
    Else
        Dim HasAlpha As Boolean
        If .Type = vbPicTypeBitmap Then
            If RenderFlag = 0 Then
                Const PICTURE_TRANSPARENT As Long = &H2
                If (.Attributes And PICTURE_TRANSPARENT) = 0 Then ' Exclude GIF
                    Dim Bmp As BITMAP
                    GetObjectAPI .Handle, LenB(Bmp), Bmp
                    If Bmp.BMBitsPixel = 32 And Bmp.BMBits <> 0 Then
                        Dim SA1D As SAFEARRAY1D, B() As Byte
                        With SA1D
                        .cDims = 1
                        .fFeatures = 0
                        .cbElements = 1
                        .cLocks = 0
                        .pvData = Bmp.BMBits
                        .Bounds.lLbound = 0
                        .Bounds.cElements = Bmp.BMWidthBytes * Bmp.BMHeight
                        End With
                        CopyMemory ByVal ArrPtr(B()), VarPtr(SA1D), 4
                        Dim i As Long, j As Long, Pos As Long
                        For i = 0 To (Abs(Bmp.BMHeight) - 1)
                            Pos = i * Bmp.BMWidthBytes
                            For j = (Pos + 3) To (Pos + Bmp.BMWidthBytes - 1) Step 4
                                If HasAlpha = False Then HasAlpha = (B(j) > 0)
                                If HasAlpha = True Then
                                    If B(j - 1) > B(j) Then
                                        HasAlpha = False
                                        i = Abs(Bmp.BMHeight) - 1
                                        Exit For
                                    ElseIf B(j - 2) > B(j) Then
                                        HasAlpha = False
                                        i = Abs(Bmp.BMHeight) - 1
                                        Exit For
                                    ElseIf B(j - 3) > B(j) Then
                                        HasAlpha = False
                                        i = Abs(Bmp.BMHeight) - 1
                                        Exit For
                                    End If
                                End If
                            Next j
                        Next i
                        CopyMemory ByVal ArrPtr(B()), 0&, 4
                    End If
                End If
                If HasAlpha = False Then RenderFlag = 1 Else RenderFlag = 2
            ElseIf RenderFlag = 2 Then
                HasAlpha = True
            End If
        End If
        If HasAlpha = False Then
            .Render hDC Or 0&, X Or 0&, Y Or 0&, CX Or 0&, CY Or 0&, 0&, .Height, .Width, -.Height, ByVal 0&
        Else
            Dim hDCBmp As Long, hBmpOld As Long
            hDCBmp = CreateCompatibleDC(0)
            If hDCBmp <> 0 Then
                hBmpOld = SelectObject(hDCBmp, .Handle)
                GdiAlphaBlend hDC, X, Y, CX, CY, hDCBmp, 0, 0, CHimetricToPixel_X(.Width), CHimetricToPixel_Y(.Height), &H1FF0000
                SelectObject hDCBmp, hBmpOld
                DeleteDC hDCBmp
            End If
        End If
    End If
End If
End With
End Sub
```

**Reply by WaynePhillipsEA** [reply]
Thanks @Kr00l, I will check this out!
